crates/typst-render/src/lib.rs

```diff
@@ -892,11 +953,51 @@ fn to_sk_paint<'a>(
 
             sk_paint.anti_alias = gradient.anti_alias();
         }
+        Paint::Pattern(pattern) => {
+            let relative = pattern.unwrap_relative(on_text);
+
+            let fill_transform = match relative {
+                RelativeTo::Self_ => fill_transform.unwrap_or_default(),
+                RelativeTo::Parent => state
+                    .container_transform
+                    .post_concat(state.transform.invert().unwrap()),
+            };
+
+            let canvas = render_pattern_frame(&state, pattern);
+            *pixmap = Some(Arc::new(canvas));
+
+            // Create the shader
+            sk_paint.shader = sk::Pattern::new(
+                pixmap.as_ref().unwrap().as_ref().as_ref(),
+                sk::SpreadMode::Repeat,
+                sk::FilterQuality::Nearest,
+                1.0,
+                fill_transform
+                    .pre_scale(1.0 / state.pixel_per_pt, 1.0 / state.pixel_per_pt),
+            );
+        }
     }
 
     sk_paint
 }
 
+fn render_pattern_frame(state: &State, pattern: &Pattern) -> sk::Pixmap {
+    let size = pattern.bbox + pattern.spacing;
+    let mut canvas = sk::Pixmap::new(
+        (size.x.to_f32() * state.pixel_per_pt).round() as u32,
+        (size.y.to_f32() * state.pixel_per_pt).round() as u32,
+    )
+    .unwrap();
+
+    canvas.fill(sk::Color::WHITE);
```

Comment:
shouldn't this be transparent rather than white? a white background can be achieved by adding a rect into the pattern, but we can't get rid of it if it's hard-coded here.